templates/nodeovc/nodeovc.py

```diff
@@ -0,0 +1,248 @@
+        for port in existent_ports:
+            if port != '22' or port not in requested_ports:
+                machine.portforward_delete(port)
+
+        # add requested ports
+        for port in port_forwards:
+            # check if ports do not exist yet
+            if port['destination'] not in existent_ports:
+                # create portforward
+                machine.portforward_create(
+                    publicport=port['destination'], 
+                    localport=port['source'], 
+                    protocol='tcp',
+                    )
+
+    def start(self):
+        """ Start the VM """
+
+        machine = self.machine
+        if machine:
+            machine.start()
+        else:
+            self.err_log_machine_not_found()
+
+    def stop(self):
+        """ Stop the VM """
+
+        machine = self.machine
+        if machine:
+            machine.stop()
+        else:
+            self.err_log_machine_not_found()
+
+    def restart(self):
+        """ Restart the VM """
+
+        machine = self.machine
+        if machine:
+            machine.restart()
+        else:
+            self.err_log_machine_not_found()
+
+    def pause(self):
+        """ Pause the VM """
+
+        machine = self.machine
+        if machine:
+            machine.pause()
+        else:
+            self.err_log_machine_not_found()
+
+    def resume(self):
+        """ Resume the VM """
+
+        machine = self.machine
+        if machine:
+            machine.resume()
+        else:
+            self.err_log_machine_not_found()
+
+    def reset(self):
+        """ Reset the VM """
+
+        machine = self.machine
+        if machine:
+            machine.reset()    
+        else:
+            self.err_log_machine_not_found()      
+
+    def snapshot(self):
+        """
+        Action that creates a snapshot of the machine
+        """
+        machine = self.machine
+        if machine:
+            machine.snapshot_create()
+        else:
+            self.err_log_machine_not_found()
+
+    def snapshot_rollback(self, snapshot_epoch=None):
+        """
+        Action that rolls back the machine to a snapshot
+        """    
+        machine = self.machine
+        if machine and snapshot_epoch:       
+            machine.snapshot_rollback(snapshot_epoch)
+            machine.start()
+        else:
+            self.err_log_machine_not_found()
+
+    def snapshot_delete(self,  snapshot_epoch=None):
+        """
+        Action that deletes a snapshot of the machine
+        """
+        machine = self.machine
+        if machine and snapshot_epoch:       
+            machine.snapshot_delete(snapshot_epoch)
+        else:
+            self.err_log_machine_not_found()        
+
+    def list_snapshots(self):
+        """
+        Action that deletes a snapshot of the machine
+        """
+        machine = self.machine
+        if machine:       
+            self.data['snapshots'] = machine.snapshots
```

Comment:
why do you store this in the data ? I think this action should return the list of snapshot directly, not storing it in the data.
And if you want to have the list of snapshot part of the data, then you need to store the snapshot in the data during the `snapshot` action remove it from the data in the `snapshot_delete`